packages/utilities/src/classNamesFunction.ts

```diff
@@ -1,32 +1,73 @@
-import { mergeStyleSets, IStyle, IStyleSet, IProcessedStyleSet } from '@uifabric/merge-styles';
+import { mergeStyleSets, IStyleSet, IProcessedStyleSet } from '@uifabric/merge-styles';
 import { IStyleFunctionOrObject } from '@uifabric/merge-styles';
 
+const MAX_CACHE_COUNT = 30;
+
+// Note that because of the caching nature within the classNames memoization,
+// I've disabled this rule to simply be able to work with any types.
+// tslint:disable:no-any
+
+const RetVal = '__retval__';
+
 /**
  * Creates a getClassNames function which calls getStyles given the props, and injects them
  * into mergeStyleSets.
+ *
+ * Note that the props you pass in on every render should be in the same order and
+ * immutable (numbers, strings, and booleans). This will allow the results to be memoized. Violating
+ * these will cause extra recalcs to occur.
  */
 export function classNamesFunction<TStyleProps extends {}, TStyleSet extends IStyleSet<TStyleSet>>(): (
   getStyles: IStyleFunctionOrObject<TStyleProps, TStyleSet> | undefined,
   styleProps?: TStyleProps
 ) => IProcessedStyleSet<TStyleSet> {
-  // TODO: memoize.
+  let map: Map<any, any> = new Map<any, any>();
 
   const getClassNames = (
     styleFunctionOrObject: IStyleFunctionOrObject<TStyleProps, TStyleSet> | undefined,
     styleProps: TStyleProps = {} as TStyleProps
   ): IProcessedStyleSet<TStyleSet> => {
-    // styleSet might be undefined if styleFunctionOrObject is undefined, but getStyles should never
-    // ordinarily be undefined (it would hardly make any sense).
-    // However, because we usually use `props.styles` as the argument to an invocation of this method, and
-    // `props.styles` itself is defined as optional, this avoids the need to use `!` at all invocation points.
-    if (styleFunctionOrObject === undefined) {
-      return {} as IProcessedStyleSet<TStyleSet>;
+    let current: Map<any, any> = map;
```

Comment:
shouldn't the template here be `Map<string, Map<string, any>>` at least, given the logic from line 33 to 47?

Reply:
In TS 3, we can further refine to `Map<string, Map<string, unknown>>`

Reply:
They key is the value of the argument, which I don't know what it is. The value is either a string or another Map<any, string | Map<any, ...etc> > recursive. Not sure how to represent.

Reply:
Oh you're right about the key. Sorry late night when I was reviewing. 🌙

Reply:
For Recursive, you can do

```
interface RecursiveMap extends Map<any, string | RecursiveMap> {    
}
```

then reference this everywhere you're using `Map<any, any>`.

Reply:
That's cool!

Reply:
I think I'm going to pass on being clever on this one.